### ISSUE TITLE
Query.onCompleted isn't fired after fetching data from the cache

### ISSUE THREAD

**Issue by elessarperm** [apollographql/react-apollo#2177]
**Intended outcome:**
I tried to make a side-effect when query result is fetched using `<Query onCompleted={...}>`

**Actual outcome:**
`onCompleted` fired after the real network request, but when I changed variables and then switched them back, query result was returned from the cache and `onCompleted` handler wasn't executed.
However, it works when I change `fetchPolicy` to `cache-and-network` or `network-only`.

**How to reproduce the issue:**
Make a `<Query>` that should log result into console when completed. Run it with some variables,
then change it to cause re-render, then return variables back to make it re-render result from cache.
`onCompleted` handler won't be fired.
fetchPolicy should be `cache-first`

**Versions**
  npmPackages:
    apollo-boost: ^0.1.10 => 0.1.10
    apollo-cache-inmemory: ^1.2.5 => 1.2.5
    apollo-client: ^2.3.5 => 2.3.5
    apollo-link: ^1.2.2 => 1.2.2
    apollo-link-http: ^1.5.4 => 1.5.4
    react-apollo: ^2.1.9 => 2.1.9

If it's not a bug, but a feature, can you explain how could I perform a side-effect after every query completion regardless of it uses cache or not?

**Comment by olistic** [comment]
I'm trying to update the state of my form with the result of a query, and I'm experiencing the same problem. `onCompleted` is not fired when the data is retrieved from the Apollo cache. This is an important issue as one cannot rely on the `onCompleted` callback if data from the cache (which plays an important role in Apollo) does not trigger it.

**Comment by Dartv** [comment]
I'm facing the same exact issue as @olistic 
Can we get this fixed?

**Comment by serranoarevalo** [comment]
Any workaround for this?

**Comment by serranoarevalo** [comment]
Adding: fetchPolicy={"cache-and-network"} to the Query Component fixed it for me.

**Comment by hxuanhung** [comment]
@serranoarevalo are you sure? I've tried but it seems doesn't solve the problem.

**Comment by serranoarevalo** [comment]
@hxuanhung yup, just tested it again and it still works. https://github.com/nomadcoders/nuber-client/blob/master/src/Routes/EditAccount/EditAccountContainer.tsx#L45

**Comment by Dartv** [comment]
It works but it's not ultimately a fix. It's a workaround at best and at a cost of extra traffic.

**Comment by laspluviosillas** [comment]
I'm having similar issues, except nothing resolves mine. onCompleted never fires regardless of what I do.

**Comment by Cabezaa** [comment]
I'm having the same problems. Has anyone been able to solve this?

**Comment by franzejr** [comment]
I got the same error. Fixed it for now by using no-cache as policy.

**Comment by laspluviosillas** [comment]
FYI, for everyone having this issue, you can work around it by managing the query manually through withApollo instead of using the Query react tag. Here's an example.

```jsx
import React from 'react';
import { withApollo } from 'react-apollo';

class Page extends React.Component {
  constructor(props) {
    super();
    this.state = {
      loading: true,
      data: {},
    };
    this.fetchData(props);
  }

  async fetchData(props) {
    const { client, productId } = props;
    const result = await client.query({
      query: YOUR_QUERY_HERE, /* other options, e.g. variables: {} */     
    });

    this.setState({
      data: result.data,
      loading: false,
    });
  }

  render() {
    const { data, loading } = this.state;
    /* 
      ... manipulate data and loading from state in here ...
    */
  }
}
export default withApollo(Page);
```

This approach relies on state so does not work with functional components, and is of course much uglier than simply <Query>, but if you *really* need a solution this should work.

**Comment by yantakus** [comment]
I can confirm that `fetchPolicy='cache-and-network'` doesn't always invoke `onCompleted` callback with this codesandbox: https://codesandbox.io/s/6zqzy802nr

And I totally agree that even if it does work in some cases, this is still a dirty hack rather than solution.

Apollo team, `onCompleted` should be always fired, that's simple logic.

**Comment by hwillson** [comment]
Confirmed - repro: https://codesandbox.io/s/50zyl4xqol
We'll get this fixed.

**Comment by Dartv** [comment]
`onCompleted` still isn't fired after mutation automatically updates cache

**Comment by hwillson** [comment]
@Dartv Can you provide a small runnable reproduction that demonstrates this?

**Comment by Dartv** [comment]
@hwillson here https://codesandbox.io/s/mm537nmqr9

**Comment by vititu** [comment]
I believe `onCompleted` is being called on every render. Before it was called once.

**Comment by tab00** [comment]
I get error "Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate." in browser console after updating to 2.2.4, preventing any rendering. The problem didn't occur in 2.1.11. Please fix this problem.

**Comment by riccoski** [comment]
I'm also getting the issue of not being called when getting it from the cache

**Comment by reyesreg** [comment]
I have two identical mutations that call the same `openSuccessBanner` in `onCompleted`.
The one that is directly passed works fine.
The one that is passed down twice fails. Such a strange error 😷

**Comment by SeanningTatum** [comment]
> Now `onCompleted` is being called on every render. Before it was called once.
> 
> That might break the app causing a `Maximum update depth exceeded` error if try to `setState` with the data provided.

I fixed it by using a `PureComponent`, because when you update the component React will re-render causing the `Query` component to run again indefinitely setting the state. 

Hope this works for you!

**Comment by vititu** [comment]
> > Now `onCompleted` is being called on every render. Before it was called once.
> > That might break the app causing a `Maximum update depth exceeded` error if try to `setState` with the data provided.
> 
> @tab00 I fixed it by using a `PureComponent`, because when you update the component React will re-render causing the `Query` component to run again indefinitely setting the state.
> 
> Hope this works for you!

@SeanningTatum I believe you are right, might be a good idea. Thanks!

**Comment by kandros** [comment]
it doesn't feel natural `onCompleted` should mean on query completed not `on read from cache` a boolean property  like onResult (which can be both from cache or network) feels more appropriated

**Comment by riccoski** [comment]
Is this working for anyone? - I'm still getting the same issue of it not being fired :(

**Comment by ridhoq** [comment]
Somehow `notifyOnNetworkStatusChange={true}` makes it work as well...which I don't understand. Is that expected behavior?

**Comment by kandros** [comment]
@ridhoq there are a couple of bugs in fetchPolicy that have unexpected behaviours when mixed with other props

**Comment by ridhoq** [comment]
@kandros hmm I understand why setting my `fetchPolicy` would cause `onCompleted` to be called, but I don't understand why `notifyOnNetworkStatusChange` would

**Comment by chitswe** [comment]
@vititu @yantakus . I use a different workaround. You have to  keep reference to completed query result with react state or instance variable. Make sure you only call `setState` only if current query result is not equal with previous keep result. Here is example.

```
<Query
onCompleted={data => {
          if ( this.lastFatchedData !== data) {
            setState({..});
          }
          this.lastFatchedData = data;
}
/>
```

**Comment by rhlsthrm** [comment]
Had the issue with the callback firing over and over, so I ended up using @laspluviosillas's solution, which worked great: https://github.com/apollographql/react-apollo/issues/2177#issuecomment-415770550

Thank you!

**Comment by Maxtermax** [comment]
with all respect apollo have really long way to be considere a stable, easy and relieble library, thar arquitecture of pass the fetch layer to a react HOC is a mistake, beacuse limit you to the react lifecycle so is  easly fell into a hole of error or glitches, currently the only safe way to set the state after a query without get hit by an error is this way:

```javascript
// CatchResponse.js
class CatchResponse extends React.Component {
  componentDidMount() {
    this.props.done(this.props.data);    
  }

  render() {
    let { children } = this.props;
    if (children) return children;
    return null;
  }
}

//Main.js
class MainComponent extends React.Component {
  state = {
    value: 0
  }
  render() {
    return <Query
       notifyOnNetworkStatusChange={true}
       query={/**/}
    >
      {
        ({ error, data, networkStatus }) => {
          let loading = !(networkStatus === 7 || networkStatus === 8);
          if (loading) return <h1>loading...</h1>
          if (data) {
            return React.cloneElement(
              <CatchResponse
                data={data}
                done={response => {
                  console.log("response ", response.name);
                  this.setState({ value: 1 });
                }}
              >
                <h1>{data.name}</h1>
              </CatchResponse>, {})
          }
        }
      }
    </Query>
  }
}
```
# Explanation

First is necessary to create a component that will wrap and render our target component, using the 
[componentDidMount](https://reactjs.org/docs/react-component.html#componentdidmount) to call a function passed by props called `done`.

```javascript
// CatchResponse.js
class CatchResponse extends React.Component {
  componentDidMount() {
    this.props.done(this.props.data);    
  }

  render() {
    let { children } = this.props;
    if (children) return children;
    return null;
  }
}
```

Be aware of the loading [issue](https://github.com/apollographql/react-apollo/issues/170#issuecomment-467716200) for that reason is necessary to add this line in the Query component

```javascript
   let loading = !(networkStatus === 7 || networkStatus === 8);
``` 

And for last use the `CatchResponse` component to receive to paint the view, note the `this.setState` inside the done function is now safe to use because with this configuration the react lifecycle is respected, also note `React.cloneElement` ensure that `CatchResponse` get remounted and the [componentDidMount](https://reactjs.org/docs/react-component.html#componentdidmount) is called and by side effect the  `this.props.done(this.props.data);`   is also called. 

```javascript
return React.cloneElement(
              <CatchResponse
                data={data}
                done={response => {
                  console.log("response ", response.name);
                  this.setState({ value: 1 });
                }}
              >
                <h1>{data.name}</h1>
              </CatchResponse>, 
       {})
             
```

**Comment by coffeedoughnuts** [comment]
FYI: I am able to recreate this issue with `"react-apollo": "2.5.5"` - `notifyOnNetworkChange` solves the problem, but obviously not ideal.

**Comment by pfarnach** [comment]
Still seeing this problem in `react-apollo: 2.5.4`.  Setting `fetch-policy="cache-and-network"` to trigger `onCompleted` works but isn't ideal.  @hwillson any updates on this?

**Comment by PavanBahuguni** [comment]
notifyOnNetworkStatusChange works and doesn't make a network call. Good workaround.

**Comment by hamxabaig** [comment]
I have also the same problem, `onCompleted` still not fires on `v2.5.6`

**Comment by garfiaslopez** [comment]
`onCompleted` still not fires on `v2.5.8`.  why is this closed?

**Comment by jasonpaulos** [comment]
@garfiaslopez or @hamxabaig, can you provide a runnable reproduction that shows this issue?

**Comment by DominicTobias** [comment]
onComplete does not fire at all for me, even though the query works :(

```
const { loading, error } = useQuery(
  gql`...`,
  {
    onComplete: data => {
      // never called :/ ???
    },
  }
);
```

**Comment by riccoski** [comment]
It's onCompleted

**Comment by hamxabaig** [comment]
@jasonpaulos here is the sandbox link where you can reproduce it https://codesandbox.io/s/pensive-worker-3py6j

Initially the query will log `fired` in console as `onCompleted` will be called. But when the query is unmounted and then mounted again, its not called anymore but the data retrieved from the cache.

`react-apollo`: `v2.5.8`

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/12872177/63020215-eb266c80-beb6-11e9-952a-cd3112ecbacc.png">

**Comment by hamxabaig** [comment]
I think it might be because of this, since neither the query nor the variables have updated, so it returns. Not sure since i'm not that aware of the codebase

https://github.com/apollographql/react-apollo/blob/9e4b573a5ae8ee068bc6f9229e6d84b36974da6b/packages/hooks/src/data/QueryData.ts#L340-L347

**Comment by jasonwr** [comment]
> @vititu @yantakus . I use a different workaround. You have to keep reference to completed query result with react state or instance variable. Make sure you only call `setState` only if current query result is not equal with previous keep result. Here is example.
> 
> ```
> <Query
> onCompleted={data => {
>           if ( this.lastFatchedData !== data) {
>             setState({..});
>           }
>           this.lastFatchedData = data;
> }
> />
> ```

This isn't a feasible approach. What if the lastFatchedData (you probably meant to say lastFetchedData) is a list (array) of objects and the data is a list of objects each 10 in length but say one of those objects has different variables. This is a shallow comparison. It's not related to this defect but this is not a feasible approach.

**Comment by jasonwr** [comment]
> I think it might be because of this, since neither the query nor the variables have updated, so it returns. Not sure since i'm not that aware of the codebase
> 
> https://github.com/apollographql/react-apollo/blob/9e4b573a5ae8ee068bc6f9229e6d84b36974da6b/packages/hooks/src/data/QueryData.ts#L340-L347

Yes but I'm getting the same exact buggy behavior with the fetchPolicy set to **no-cache** This is a bug.

**Comment by hwillson** [comment]
@jasonwr If this is still happening with React Apollo 3.0.1, would you mind opening a new issue about this (ideally with a small runnable reproduction)? Thanks!

**Comment by jasonwr** [comment]
> @jasonwr If this is still happening with React Apollo 3.0.1, would you mind opening a new issue about this (ideally with a small runnable reproduction)? Thanks!

@hwillson turns out this might be a user error. I was on 2.5. Just submitted my code to review but the review is going to have to wait for some other things to get through the pipeline. I'll keep you all posted. Thanks.

**Comment by websash** [comment]
@hwillson The onCompleted callback is still not working correctly in @apollo/react-hooks 3.0.1 It runs on every component rerender when I change its state. Not expected behavior for this callback. I've created a codesandbox https://codesandbox.io/s/jolly-easley-ekm0j 
I can open a new issue if there's not one open already

**Comment by cristiangrojas** [comment]
I can't really believe this is still happening and there's no fix.

**Comment by dylanwulf** [comment]
@websash Check out #3353. That problem has been fixed, but hasn't been released yet. The fix will be included in the next release

**Comment by tresdev** [comment]
Is there any other option I can use instead of onCompleted, some funciton I can use after the query result is retrieved whether is done on network or from cache.

**Comment by fnune** [comment]
Included here too https://github.com/apollographql/react-apollo/issues/3535 is the related issue that `Query.onCompleted` only fires for the initial fetch if using `pollInterval`. All subsequent polling requests fail to call `Query.onCompleted`.

**Comment by Algram** [comment]
@tresdev we solved it by using the useEffect react hook to react to a completed query:

```
  const { data, loading, error } = useQuery(queryName, {
    pollInterval: 15000,
  })

  useEffect(() => {
    if (loading || error) {
      return
    }

    // Do your thing with the data here
  }, [data, error, loading])
```

**Comment by limekiln** [comment]
@Algram this does not work for me since none of the variables are changing on the second query (data, loading, error, networkStatus). Is there another workaround?

**Comment by fgiarritiello** [comment]
Please, any chance we can get this fixed?

**Comment by SamiUrias** [comment]
It still not working, why this issue gets closed?

**Comment by henryhobhouse** [comment]
Also having this issue with useLazyQuery. 

* Make first request (network). onCompleted is firing. Correct data returned
* Make second request (network). onCompleted is firing.  Correct data returned
* Make first request again (cache). onCompleted is **not** firing.  Correct data returned

**Comment by martdavidson** [comment]
This is still happening, can we get this reopened?

**Comment by dqunbp** [comment]
> This is still happening, can we get this reopened?

I acknowledge this issue.
Using `cache-and-network` leads to overuse of the network.
Maybe we need an additional method, for example `onCompletedFromCache`.

**Comment by dryoma** [comment]
> Maybe we need an additional callback, for example `onCompletedFromCache`?

I'd suggest having one callback for all "completed" events, but pass some argument with the info on how exactly the data have been fetched: from cache or via network.

**Comment by Darklaki** [comment]
I'm facing the same issue.
Having one callback and passing an argument whether to fire onCompleted twice (when we get data form cache and then when we receive data from server) would work like a charm.

**Comment by PavanBahuguni** [comment]
use notifyOnNetworkChange when u are using default fetch-policy (when u havent specified anythig). @Darklaki

**Comment by palerdot** [comment]
This issue is a dealbreaker and kind of makes the default `fetchPolicy: cache-first` useless in `useQuery` ... With this bug it is impossible to prevent subsequent expensive api calls, since the UI is not notified of data from cache ... 

Please reopen this issue as it is a significant bug forcing UI to make graphql calls everytime, even though UI could do fine with results from few seconds back for the same set of input.

**Comment by PavanBahuguni** [comment]
@palerdot I have beein using notifyOnNetworkChange on Query render prop, it does call the onComleted when that data is fetched from network or cache, u dont have to change you'r network policy.

**Comment by palerdot** [comment]
@PavanBahuguni It does not seem to work for me ... plus I'm using `useQuery` and it is not working for `useQuery` hook

**Comment by mvirbicianskas** [comment]
I can confirm, this is not working for me too in @apollo/react-hooks@3.1.3

**Comment by himat** [comment]
Same as @mvirbicianskas If I try to fetch the same data consecutively, it doesn't work

**Comment by dominik-myszkowski** [comment]
fetchPolicy: 'network' option fixes it successfully. onCompleted fires every time

**Comment by fnune** [comment]
> fetchPolicy: 'network' option fixes it successfully. onCompleted fires every time

It doesn't fix it: it circumvents the issue by not using the cache. This is about `onCompleted` not firing when the data comes from the cache.

**Comment by Bolza** [comment]
It has been 2 years this issues hasn't been fixed and the issue has been closed.
It seems a pretty important bug, it's easy to reproduce and there's a very high chance to end up in this situation. Can anyone explain why it's closed?

**Comment by cyberwombat** [comment]
Reopened here https://github.com/apollographql/react-apollo/issues/3968 - hopefully it will get some visibility. Pretty much a show stopper for me.

**Comment by 1awaleed** [comment]
I can't believe this is still not fixed.

**Comment by farid-ouachrar** [comment]
Any update regarding this issue?

**Comment by vitrukroman** [comment]
fetch policy `cache-and-network` is not correct overcome even
Imagine you have loading spinner which hides onCompleted . 
When data is cached, it is immediately returned, however spinner will be loading until network request is in progress.

I agree that onCompleted event must be called when data is retrieved from cache

**Comment by fadomire** [comment]
almost 2 years and no explanation of the why it is like that or a fix
seems like it is time to switch from apollo
